test/unit/mongo_client.test.js

```diff
@@ -847,4 +848,68 @@ describe('MongoOptions', function () {
       });
     });
   });
+
+  context('MongoLogger', function () {
```

Comment:
as I mentioned in the other file, the job of the options parsing is to ensure that we are correctly parsing the options into the format expected by the logging class, so I would expect these tests to be updated to test that the options are set as expected;

I would actually go a step further and outsource the instantiation of the logger from the options to the logger class itself (i.e., let the logger class constructor return `null`), that way we have a clear division of responsibilities: the options parser consolidates options, the logger class decides what to do with them

Reply:
Can you elaborate on this?  The logger class constructor currently returns null.  With the exception of one option (that could be moved to resolveOptions), there is already a clear separation of responsibilities?

These tests are responsible for testing whether or not the actual logger gets instantiated on the MongoClient, not anything related to options parsing.

Reply:
this context `MongoLogger` is nested inside the `MongoOptions` context, but `mongoLogger` should not be a member of `MongoOptions` at all, we should have the **parsed** options be saved onto the `MongoOptions` object, and this file should be updated to only test that these standardized options are correctly assigned to the options object (we can even keep `mongoLogger` but make it reference the options passed to it instead of the actual instance of the logger)

the options parsing should **not** generate the logger;
the client should be updated to have:
```
// instead of this: this.mongoLogger = this[kOptions].mongoLogger ?? null;

// do this:
this.mongoLogger = new MongoLogger(<this[kOptions].mongoLoggerOptions>);

// or this:
this.mongoLogger = MongoLogger.create(<this[kOptions].mongoLoggerOptions>);
// should return an instance of the logger if some options are on, otherwise can return null
```

with that change, this file should only test that the options are correctly assigned to the client options object
and the mongo logger test should make sure that all combinations of options result in the correct return result from `new MongoLogger()` or `MongoLogger.create()`

The problem with the current implementation is that we don't preserve what options are used to configure the logger, which is suboptimal for debugging, and we have the logic of the resulting logging scattered across 3 places: options parsing, client constructor, and the logger class itself. The job of options parsing should be only to validate the options and transform them into a standardized form. The job of the client constructor should be to initialize whatever APIs it needs to have. The job of the MongoLogger class should be to determine what to do with the options, including what to do when all options are off.

The only reason this comment is in the testing section is because once we do have the options saved on the options object, we'll want to make sure that they are tested here. If we just return the result of the logger option parser, then we can also just use sinon to make sure that that's what we assign to the corresponding client options.

Reply:
bailey beat me to it, but thanks for elaborating. really helped me clear up a lot the separation of responsibilities. unfortunately no relevant code changes because of our team conversation from slack where we agreed that `mongoLogger` should always be present :(

Reply:
yes, there are changes - we should add tests to make sure that the expected `mongoLoggerOptions` are getting set on the options attached to the client